### ISSUE TITLE
DNS resolver as an extension point

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#14310]
We currently have c-ares and Apple iOS DNS resolver support as compile time options. We might also be interested in a site-specific DNS resolver library. I think it's reasonable to make this a standard extension point.

@junr03 since you were last digging into this, any thoughts on whether it would be feasible to accommodate the various DNS resolvers this way?

CC @yanavlasov @antoniovicente @asraa @adisuissa @KBaichoo @akonradi

**Comment by junr03** [comment]
It would certainly be feasible. @mattklein123 and I discussed it somewhere  (I am trying to find where) when I was adding the Apple Resolver, and decided for the current setup for expediency. However, there aren't any **major** technical reasons why the resolvers couldn't be formalized as an extension point. In fact, it would help clean up configuration that doesn't apply to all resolvers.

Resolvers are created via the dispatcher, so there is already a small surface for resolver creation. In terms of configuration consumption there are a few places where the extension would have to be accessible: some clusters, dns cache, udp dns filter; and additionally the "default" server resolver.

So overall, my assessment is that while there is some cleaning up to do it would actually be nice to elevate resolvers as an extension point, and something I would have wanted to do if I had had more time in the past. Happy to help whomever wants to take this on.

**Comment by mattklein123** [comment]
Yes I agree making the DNS resolver a full extension would be good.

**Comment by yanjunxiang-google** [comment]
I am looking into this issue now.

**Comment by yanjunxiang-google** [comment]
@yanavlasov

**Comment by htuch** [comment]
Reopening, since #17272 was just the API changes (@yanjunxiang-google best to put "Part of #xyzabc" instead of "Fixes #xyzabc" when working on a series of patches on an issue).
